tasks/purgeChannel.js

```diff
@@ -0,0 +1,151 @@
+    message.createdAt < (new Date() - maxAgeInMilliSeconds)
+}
+
+const checkBulk = (message, maxAge) => {
+  const maxAgeInMilliSeconds = maxAge * 60 * 60 * 1000
+  return !message.pinned &&
+    message.createdAt > (new Date() - twoWeeksInMilliSeconds) &&
+    message.createdAt < (new Date() - maxAgeInMilliSeconds)
+}
+
+const checkNonBulk = (message, maxAge) => {
+  const maxAgeInMilliSeconds = maxAge * 60 * 60 * 1000
+  return !message.pinned &&
+    message.createdAt < (new Date() - twoWeeksInMilliSeconds) &&
+    message.createdAt < (new Date() - maxAgeInMilliSeconds)
+}
+
+const initPurgeChannelTasksAsync = async (client) => {
+  _client = client
+  const guildsData = await getAllGuilds()
+  guildsData.forEach(guildData => {
+    Object.entries(guildData.purgers).forEach(([channelId, purger]) => {
+      if (!tasks[channelId]) {
+        addPurgerAndStartTask(purger)
+      }
+    })
+  })
+}
+
+const removePurgeChannelTask = (purger) => {
+  if (!tasks[purger.channelId]) {
+    return
+  }
+  tasks[purger.channelId].stop()
+  delete tasks[purger.channelId]
+  log.info(`stopped and removed Purge task on channel <#${purger.channelId}>. Its settings were: max ${purger.maxAge} hours - cronTime: ${purger.cronTime}`)
+}
+
+const addPurgerAndStartTask = (purger) => {
+  tasks[purger.channelId] = cron.job(
+    purger.cronTime,
+    async () => {
+      try {
+        const channelToPurge = await _client.channels.cache.find(channel => channel.id === purger.channelId)?.fetch()
+        if (!channelToPurge) {
+          log.warn(`Task Purge failed: channel <#${purger.channelId}> does not exist`)
+          return
+        }
+        log.info(`Task Purge: channel ${channelToPurge.name} <#${purger.channelId}>`)
+
+        let lastFetchedMessageId // can be undefined for first fetch
+        let messagesToDeleteIndividually = new Collection()
+        let countDeletedMessage = 0
+        // first use bulkDelete to reduce API calls, but bulkDelete doesn't delete messages older than 2 weeks (discord API limit)
+        while (true) {
+          // get last messages, starting at our last fetched message, or undefined if it's the first time we run this
+          const fetchedMessages = await channelToPurge.messages.fetch({ limit, before: lastFetchedMessageId })
+
+          // if there are no fetched messages, stop
+          if (fetchedMessages.size === 0) {
+            break
+          }
+          lastFetchedMessageId = fetchedMessages.last().id
+
+          // filter for bulk removal
+          const messagesToBulkDelete = fetchedMessages.filter(message => checkBulk(message, purger.maxAge))
+
+          // add non bulk removable message to messagesToDelete
+          messagesToDeleteIndividually = messagesToDeleteIndividually.concat(fetchedMessages.filter(message => checkNonBulk(message, purger.maxAge)))
+
+          // actually delete them
+          const deletedMessages = await channelToPurge.bulkDelete(messagesToBulkDelete)
```

Comment:
Does the bulk delete also limit the max amount of messages like the fetch does?
If not, wouldn't it be more efficient to delete them all together after the loop?

Reply:
i believe it also has the limit